lib/Sema/CSApply.cpp

```diff
@@ -5451,12 +5447,201 @@ static bool hasCurriedSelf(ConstraintSystem &cs, ConcreteDeclRef callee,
+  unsigned paramIdx;
+  for (paramIdx = 0; paramIdx != parameterBindings.size(); ++paramIdx) {
+    if (llvm::find(parameterBindings[paramIdx], *unlabeledTrailingClosureIndex)
+          != parameterBindings[paramIdx].end())
+      break;
+  }
+  assert(paramIdx != parameterBindings.size() && "Unbound trailing closure!");
+
+  // If this parameter requires an argument, it would have been unfilled
+  // prior to SE-2086; there is nothing to diagnose.
+  if (parameterRequiresArgument(params, paramInfo, paramIdx))
+    return;
+
+  // Look for a later parameter that could match a trailing closure; the
+  // last one of these would have been picked prior to SE-0286.
+  Optional<unsigned> matchingBackwardParamIdx;
+  for (unsigned backwardParamIdx :
+           range(paramIdx + 1, parameterBindings.size())) {
+    if (!paramInfo.acceptsUnlabeledTrailingClosureArgument(backwardParamIdx))
+      continue;
+
+    matchingBackwardParamIdx = backwardParamIdx;
+  }
+
+  // If there is no other parameter that could match the unlabeled trailing
+  // closure, there is nothing to diagnose.
+  if (!matchingBackwardParamIdx)
+    return;
+
+  // Do a simple arity check; if the matched parameter and backward-matched
+  // parameter accept functions with different arity, this would not have
+  // type-checked with the backward scan, so there is nothing to report.
+  if (functionParameterArity(params[paramIdx]) !=
+        functionParameterArity(params[*matchingBackwardParamIdx]))
+    return;
+
+  // Compute the various source locations where diagnostics will occur.
+  ASTContext &ctx = params[paramIdx].getPlainType()->getASTContext();
+  Expr *trailingClosure;
+  SourceLoc existingRParenLoc;
+  SourceLoc leadingCommaLoc;
+  if (auto tupleExpr = dyn_cast<TupleExpr>(arg)) {
+    trailingClosure = tupleExpr->getElements().back();
+    existingRParenLoc = tupleExpr->getRParenLoc();
+    assert(tupleExpr->getNumElements() >= 2 && "Should be a ParenExpr?");
+    leadingCommaLoc = Lexer::getLocForEndOfToken(
+        ctx.SourceMgr,
+        tupleExpr->getElements()[tupleExpr->getNumElements()-2]->getEndLoc());
+  } else {
+    auto parenExpr = cast<ParenExpr>(arg);
+    trailingClosure = parenExpr->getSubExpr();
+    existingRParenLoc = parenExpr->getRParenLoc();
+  }
+
+  // Determine the names of the parameters that would be matched by the
+  // forward and backward scans.
+  Identifier paramName = params[paramIdx].getLabel();
+  Identifier backwardParamName = params[*matchingBackwardParamIdx].getLabel();
+
+  // Produce a diagnostic referencing the callee.
+  auto noteCallee = [&] {
+    auto decl = callee.getDecl();
+    if (!decl)
+      return;
+
+    auto diag = ctx.Diags.diagnose(
+        decl, diag::decl_multiple_defaulted_closure_parameters,
+        decl->getName(), paramName, backwardParamName);
+
+    // Dig out the parameter declarations so we can highlight them.
+    // FIXME: There should be a utility for this.
```

Comment:
You can use `swift::getParameterList` :)

Reply:
Fantastic, thank you! I've switched over to using that here.